uefi/src/proto/scsi/pass_thru.rs

```diff
@@ -0,0 +1,239 @@
+    /// - `scsi_req`: The [`ScsiRequest`] containing the command and data to send to the device.
+    ///
+    /// # Returns
+    /// [`ScsiResponse`] containing the results of the operation, such as data and status.
+    ///
+    /// # Errors
+    /// - [`Status::SUCCESS`] The SCSI Request Packet was successfully sent. For bi-directional
+    ///   commands, `InTransferLength` bytes were transferred from `InDataBuffer`. For write and
+    ///   bi-directional commands, `OutTransferLength` bytes were transferred by `OutDataBuffer`.
+    ///   Additional status information is available in `HostAdapterStatus`, `TargetStatus`,
+    ///   `SenseDataLength`, and `SenseData`.
+    /// - [`Status::BAD_BUFFER_SIZE`] The SCSI Request Packet was not executed because the data
+    ///   buffer size exceeded the allowed transfer size for a single command. The number of bytes
+    ///   that could be transferred is returned in `InTransferLength` or `OutTransferLength`.
+    /// - [`Status::NOT_READY`] The SCSI Request Packet could not be sent because too many packets
+    ///   are already queued. The caller may retry later.
+    /// - [`Status::DEVICE_ERROR`] A device error occurred while attempting to send the SCSI Request Packet.
+    ///   Additional status information is available in `HostAdapterStatus`, `TargetStatus`, `SenseDataLength`,
+    ///   and `SenseData`.
+    /// - [`Status::INVALID_PARAMETER`] The `Target`, `Lun`, or the contents of `ScsiRequestPacket` are invalid.
+    ///   The SCSI Request Packet was not sent, and no additional status information is available.
+    /// - [`Status::UNSUPPORTED`] The command described by the SCSI Request Packet is not supported by the
+    ///   host adapter, including unsupported bi-directional SCSI commands. The SCSI Request Packet was not
+    ///   sent, and no additional status information is available.
+    /// - [`Status::TIMEOUT`] A timeout occurred while executing the SCSI Request Packet. Additional status
+    ///   information is available in `HostAdapterStatus`, `TargetStatus`, `SenseDataLength`, and `SenseData`.
+    pub fn execute_command<'req>(
+        &self,
+        mut scsi_req: ScsiRequest<'req>,
+    ) -> crate::Result<ScsiResponse<'req>> {
+        unsafe {
+            (self.proto.pass_thru)(
+                self.proto,
+                self.target_lun.0.as_ptr(),
+                self.target_lun.1,
+                &mut scsi_req.packet,
+                ptr::null_mut(),
+            )
+            .to_result_with_val(|| ScsiResponse(scsi_req))
+        }
+    }
+}
+
+/// An iterator over SCSI devices available on the channel.
+#[derive(Debug)]
+pub struct ScsiTargetLunIterator<'a> {
+    proto: &'a ExtScsiPassThruProtocol,
+    prev: ScsiTargetLun,
+}
+impl<'a> Iterator for ScsiTargetLunIterator<'a> {
+    type Item = ScsiDevice<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // no idea why get_next_target_lun() takes a doubleptr ... makes no sense
```

Comment:
See spec :)
![image](https://github.com/user-attachments/assets/0f5199df-4118-41b7-90ac-a75d9f7e4a06)

Reply:
Yeah, I read that when implementing it. But my testing showed, that the call never returned a new pointer, it instead always wrote the response into the target buffer I gave it. And for that -  a double pointer doesn't make sense.

But now that I think about it ... the spec formulation probably means that the UEFI implementation is **allowed** to return me a new buffer instead of overwriting my buffer. I'll fix my implementation to honor that.